File: python/src/laakhay/ta/indicators/trend/hma.py
"""Hull Moving Average (HMA) indicator implementation."""

from __future__ import annotations

import ta_py

from ...core import Series
from ...core.types import Price
from ...registry.models import SeriesContext
from ...registry.registry import register
from ...registry.schemas import (
    IndicatorSpec,
    OutputSpec,
    ParamSpec,
    RuntimeBindingSpec,
    SemanticsSpec,
)
from .._utils import results_to_series

HMA_SPEC = IndicatorSpec(
    name="hma",
    description="Hull Moving Average (fast, lag-reduced)",
    params={"period": ParamSpec(name="period", type=int, default=14, required=False)},
    outputs={"result": OutputSpec(name="result", type=Series, description="HMA values", role="line")},
    semantics=SemanticsSpec(
        required_fields=("close",),
        lookback_params=("period",),
    ),
    runtime_binding=RuntimeBindingSpec(kernel_id="hma"),
)


@register(spec=HMA_SPEC)
def hma(ctx: SeriesContext, period: int = 14) -> Series[Price]:
    """Hull Moving Average (HMA) implementation."""
    if period <= 0:
        raise ValueError("HMA period must be positive")
    result = ta_py.hma([float(v) for v in ctx.close.values], period)
    return results_to_series(result, ctx.close, value_class=Price)

File: crates/ta-engine/src/core/metadata/trend/hma.rs
use super::*;

pub const META: IndicatorMeta = IndicatorMeta {
    id: "hma",
    display_name: "Hull Moving Average",
    category: "trend",
    aliases: &[],
    param_aliases: &[PARAM_ALIAS_LOOKBACK_PERIOD],
    params: &[P_PERIOD_14],
    outputs: &[IndicatorOutputMeta {
        name: "result",
        kind: "line",
        description: "HMA value",
    }],
    semantics: SEM_CLOSE_PERIOD,
    visual: VIS_PRICE_LINE,
    runtime_binding: "hma",
};
